### ISSUE TITLE
Binary buffer saved incorrectly

### ISSUE THREAD

**Issue by ValYouW** [Automattic/mongoose#5551]
This is a reopen of #5497 (as I couldn't get a response there).
Reproduction was on Mongoose v4.11.7

Calling `findOneAndUpdate` and updating a binary field saves the binary data incorrectly.
As mentioned in #5497 this is a breaking change that apparently introduced in version 4.11.2 (I suspect this commit: 1613ef5).

Please see the reproduction script below (which include `runSettersOnQuery: true`):

```js
var mongoose = require('mongoose'),
	uuid = require('uuid'),
	uuidParse = require('uuid-parse');

mongoose.Promise = Promise;

function generateUUID() {
	var buffer = uuid.v4(null, new Buffer(16));
	return new mongoose.Types.Buffer(buffer).toObject(0x04);
}

function toUUID(string) {
	if (!string) {return null;}
	if (Buffer.isBuffer(string) || Buffer.isBuffer(string.buffer)) { return string; }
	var buffer = uuidParse.parse(string);
	return new mongoose.Types.Buffer(buffer).toObject(0x04);
}

function fromUUID(buffer) {
	if (!buffer || buffer.length !== 16) {return null;}
	return uuidParse.unparse(buffer);
}

var UserSchema = new mongoose.Schema({
	name: String,
	foo: {
		type: mongoose.Schema.Types.Buffer,
		get: fromUUID,
		set: toUUID
	}
}, {collection: 'users'});

var UserModel = mongoose.model('User', UserSchema);

var connStr = 'mongodb://test:test@127.0.0.1/test';
mongoose.connect(connStr);

var user = {name: 'insert', foo: uuid.v4()};
console.log('Inserting user, my foo will be OK:', user);
var dalUser = new UserModel(user);
dalUser.save().then(() => {
	console.log('User inserted successfully');
}).catch(console.error);


user = {name: 'upsert', foo: uuid.v4()};
console.log('Upserting user, my foo will be wrong:', user);
var opts = {upsert: true, setDefaultsOnInsert: true, runSettersOnQuery: true };
UserModel.findOneAndUpdate({foo: user.foo}, user, opts).exec().then(() => {
	console.log('User upserted successfully');
}).catch(console.error);
```

And the result in the DB is:
![image](https://user-images.githubusercontent.com/4539636/29287951-a9c72632-813f-11e7-9d8c-2c3a24319092.png)

Thanks!

**Comment by vkarpov15** [comment]
Will double check why this changed, thanks for your patience.

**Comment by vkarpov15** [comment]
Thanks for your patience @ValYouW and thanks for the very detailed bug report. We didn't notice it because the mongoose logs showed that we were saving with `sub_type = 4`, but underneath the logs it ended up becoming `sub_type = 0`. Fix will be in 4.11.9 tomorrow :+1: :beers:
